### PR TITLE
Make the HttpTraceContext be more forgiving in the face of invalid data.

### LINKED ISSUE #895: BUG: Invalid trace context header crashes extract
See e.g.: https://github.com/open-telemetry/opentelemetry-java/blob/f4dea297be6eac28eade6dd1895ad73eed28d8f8/api/src/main/java/io/opentelemetry/trace/propagation/HttpTraceContext.java#L137

This is not in compliance with https://github.com/open-telemetry/opentelemetry-specification/blob/69768aa7bb8bdcd2c1490e5a1f38e109dc0517a3/specification/error-handling.md#basic-error-handling-principles:

> OpenTelemetry implementations MUST NOT throw unhandled exceptions at run time.

https://github.com/open-telemetry/opentelemetry-specification/issues/478 is related to this, but this issue here also concerns traceparent.

An invalid (non-null) spancontext should be returned instead of throwing.

### PR DESCRIPTION
**Pull request by jkwatson**
resolves #895 

Note: This does not fully address #876 . That will wait until https://github.com/open-telemetry/opentelemetry-specification/issues/478 is resolved.

### CODE REVIEW

**Comment by carlosalberto** [comment]
Looks great overall, a rather needed clean up, thanks for working on it. Don't quite like the potential creation of two `SpanContext` instances though ;)

**Comment by jkwatson** [comment]
> Looks great overall, a rather needed clean up, thanks for working on it. Don't quite like the potential creation of two `SpanContext` instances though ;)

I'd rather easy-to-understand code that we can optimize *if necessary* than having hard-to-understand code to start with.

File: api/src/main/java/io/opentelemetry/trace/propagation/HttpTraceContext.java [hunk 1]
```diff
@@ -111,53 +114,66 @@
...
+    // TODO(bdrutu): Do we need to verify that version is hex and that
+    // for the version the length is the expected one?
+    checkArgument(
+        traceparent.charAt(TRACE_OPTION_OFFSET - 1) == TRACEPARENT_DELIMITER
+            && (traceparent.length() == TRACEPARENT_HEADER_SIZE
+                || (traceparent.length() > TRACEPARENT_HEADER_SIZE
+                    && traceparent.charAt(TRACEPARENT_HEADER_SIZE) == TRACEPARENT_DELIMITER))
+            && traceparent.charAt(SPAN_ID_OFFSET - 1) == TRACEPARENT_DELIMITER
+            && traceparent.charAt(TRACE_OPTION_OFFSET - 1) == TRACEPARENT_DELIMITER,
+        "Missing or malformed TRACEPARENT.");
```
Comment (bogdandrutu) [inline api/src/main/java/io/opentelemetry/trace/propagation/HttpTraceContext.java]: If we always return Invalid why do we create exceptions, maybe just simply return Invalid from here.

Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/trace/propagation/HttpTraceContext.java]: It's a great question. And a great answer. I'll make it so.  :)

Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/trace/propagation/HttpTraceContext.java]: cleaned this up a bit to not use exceptions where not currently necessary. We could get rid of some more exception handling by adding methods on the TraceId, SpanId to do format validation without exceptions.

File: api/src/main/java/io/opentelemetry/trace/propagation/HttpTraceContext.java [hunk 2]
```diff
@@ -111,53 +114,73 @@
   public <C /*>>> extends @NonNull Object*/> SpanContext extract(C carrier, Getter<C> getter) {
     checkNotNull(carrier, "carrier");
```
Comment (Oberon00) [inline api/src/main/java/io/opentelemetry/trace/propagation/HttpTraceContext.java]: What about these? Especially a `null` carrier might make sense if I use a lambda or anonymous inner class as getter that already holds references to the actual carrier.

Comment (thisthat) [reply api/src/main/java/io/opentelemetry/trace/propagation/HttpTraceContext.java]: I think ```extract(getter.getCarrier(), getter)``` is more clean than using `null` here and there

Comment (Oberon00) [reply api/src/main/java/io/opentelemetry/trace/propagation/HttpTraceContext.java]: I was thinking about:
```java
HttpResponse resp = ...;
SpanContext sctx = format.extract(null, (c, key) -> resp.get(key))
```
vs
```java
HttpResponse resp = ...;
SpanContext sctx = format.extract(resp, (c, key) -> c.get(key))
```

I think both are equally valid, and in any case like mentioned in #895, the API/SDK must not use exceptions for error reporting.

Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/trace/propagation/HttpTraceContext.java]: I agree this is a valid way to call things. But, let's do that as a separate change; please open an issue about it if you feel we need to change this API to support it.

**Pull request merged**